core/executor/src/wasm_executor.rs

```diff
@@ -638,74 +636,109 @@ impl WasmExecutor {
 			// these assumptions are probably not true anymore
 			.expect("all modules compiled with rustc should have an export named 'memory'; qed")
 			.as_memory()
 			.expect("in module generated by rustc export named 'memory' should be a memory; qed")
 			.clone();
-		let table: Option<TableRef> = intermediate_instance
-			.not_started_instance()
+		let table: Option<TableRef> = module_instance
 			.export_by_name("__indirect_function_table")
 			.and_then(|e| e.as_table().cloned());
 
-		let mut fec = FunctionExecutor::new(memory.clone(), heap_pages, table, ext)?;
-
-		// finish instantiation by running 'start' function (if any).
-		let instance = intermediate_instance.run_start(&mut fec)?;
+		let low = memory.lowest_used();
+		let used_mem = memory.used_size();
+		let mut fec = FunctionExecutor::new(memory.clone(), table, ext)?;
 		let size = data.len() as u32;
 		let offset = fec.heap.allocate(size);
 		memory.set(offset, &data)?;
 
-		let result = instance.invoke_export(
+		let result = module_instance.invoke_export(
 			method,
 			&[
 				I32(offset as i32),
 				I32(size as i32)
 			],
 			&mut fec
 		);
-		let returned = match result {
-			Ok(x) => x,
+		let result = match result {
+			Ok(Some(I64(r))) => {
+				let offset = r as u32;
+				let length = (r >> 32) as u32 as usize;
+				memory.get(offset, length)
+					.map_err(|_| ErrorKind::Runtime.into())
+			},
+			Ok(_) => Err(ErrorKind::InvalidReturn.into()),
 			Err(e) => {
-				trace!(target: "wasm-executor", "Failed to execute code with {} pages", heap_pages);
-				return Err(e.into())
+				trace!(target: "wasm-executor", "Failed to execute code with {} pages", memory.current_size().0);
+				Err(e.into())
 			},
 		};
 
-		if let Some(I64(r)) = returned {
-			let offset = r as u32;
-			let length = (r >> 32) as u32 as usize;
-			memory.get(offset, length)
-				.map_err(|_| ErrorKind::Runtime.into())
-		} else {
-			Err(ErrorKind::InvalidReturn.into())
+		// cleanup module instance for next use
+		let new_low = memory.lowest_used();
+		if new_low < low {
+			memory.zero(new_low as usize, (low - new_low) as usize)?;
+			memory.reset_lowest_used(low);
 		}
+		memory.with_direct_access_mut(|buf| buf.resize(used_mem.0, 0));
+		result
+	}
+
+	/// Prepare module instance
+	pub fn prepare_module<E: Externalities<Blake2Hasher>>(
+		&self,
+		ext: &mut E,
+		heap_pages: usize,
+		module: &Module,
+		) -> Result<ModuleRef>
+	{
+		// start module instantiation. Don't run 'start' function yet.
+		let intermediate_instance = ModuleInstance::new(
+			module,
+			&ImportsBuilder::new()
+			.with_resolver("env", FunctionExecutor::<E>::resolver())
+			)?;
+
+		// extract a reference to a linear memory, optional reference to a table
+		// and then initialize FunctionExecutor.
+		let memory = intermediate_instance
+			.not_started_instance()
+			.export_by_name("memory")
+			// TODO: with code coming from the blockchain it isn't strictly been compiled with rustc anymore.
```

Comment:
No TODOs without links to issues in PRs to master: please put the info into a ticket and provide the link here or remove the TODO :) .

Reply:
This TODO is copy-pasted from a similar call above. I'm not sure what it is about to be honest.

Reply:
AFAIR, initially we had an `expect("this code is from rustc ; it's always correct")`, but then somebody actually noted that this code isn't actually coming from a rustc generated wasm artifact but from the blockchain storage and so it became harder to claim that rustc generated this binary.

I'm not sure if we want to panic here, it would be better if we could return an `Err` here.

Reply:
@pepyakin maybe you could put up a ticket for it that we can link here from? I feel the change is outside of the scope of this PR but the criticism correct and should be resolved (IMHO with an `Err`)

Reply:
Returns error now